### ISSUE TITLE
Static source code analysis report

### ISSUE THREAD

**Issue by rodrigoffzz** [lwfinger/rtl8723ds#34]
Hi,

I'm working with the Ubuntu security team and we were asked to give this project a very quick review as part of the Main Inclusion Review[1].
Part of the review is to run Coverity, a static source code analyzer. The outputs are interesting and useful from a security perspective.
In case it is useful, I'm attaching here the full report from Coverity scan in a text based output: [coverity.txt](https://github.com/lwfinger/rtl8723ds/files/10539860/coverity.txt). If you are interested, you can sign up for a free account in https://scan.coverity.com and it will give you a better experience viewing the issues.

Some of them are worth looking and some I wonder if it is really an issue or not.

For example, I've selected the following report, for a uninitialized variable[2]:
```
core/rtw_ieee80211.c:1413:2:
  Type: Uninitialized scalar variable (UNINIT)

core/rtw_ieee80211.c:1383:2:
  1. var_decl: Declaring variable "mac" without initializer.
core/rtw_ieee80211.c:1385:2:
  2. path: Condition "out == NULL", taking false branch.
core/rtw_ieee80211.c:1391:2:
  3. path: Condition "rtw_initmac", taking false branch.
core/rtw_ieee80211.c:1407:2:
  4. path: Condition "hw_mac_addr", taking false branch.
core/rtw_ieee80211.c:1413:2:
  5. uninit_use_in_call: Using uninitialized value "*mac" when calling "rtw_check_invalid_mac_address".
core/rtw_ieee80211.c:1348:2:
  5.1. read_value: Reading value "*mac_addr" when calling "_rtw_memcmp".
os_dep/osdep_service.c:721:2:
  5.1.1. read_value: Reading value "*dst" when calling "memcmp". [Note: The source code implementation of the function has been overridden by a builtin model.]
os_dep/osdep_service.c:721:2:
  5.1.2. path: Condition "!memcmp(dst, src, sz)", taking true branch.
core/rtw_ieee80211.c:1348:2:
  5.2. path: Condition "_rtw_memcmp(mac_addr, null_mac_addr, 6)", taking true branch.
core/rtw_ieee80211.c:1350:3:
  5.3. path: Jumping to label "func_exit".
```
* We can, for example, reach the case where `mac_addr` has a random value from the memory and the driver may be lead to an undefined behavior.

In the other hand, we have some false-positives, like below:
```
os_dep/linux/ioctl_linux.c:10206:4:
  Type: Out-of-bounds write (OVERRUN)

os_dep/linux/ioctl_linux.c:9593:2:
  1. path: Condition "copy_from_user(extra, wrqu->pointer, wrqu->length)", taking false branch.
os_dep/linux/ioctl_linux.c:9599:2:
  2. path: Condition "setdata == NULL", taking false branch.
...
os_dep/linux/ioctl_linux.c:10179:3:
  21. path: Condition "hal_efuse_macaddr_offset(padapter) == -1", taking false branch.
os_dep/linux/ioctl_linux.c:10184:3:
  22. return_constant: Function call "hal_efuse_macaddr_offset(padapter)" may return -1.
os_dep/linux/ioctl_linux.c:10184:3:
  23. assignment: Assigning: "addr" = "hal_efuse_macaddr_offset(padapter)". The value of "addr" is now 65535.
...
  29. overrun-local: Overrunning array "pEfuseHal->fakeEfuseModifiedMap" of 1024 bytes at byte offset 65535 using index "addr + jj" (which evaluates to 65535).
```
* In this case, Coverity is not taking into account that, in Condition 21, there is an error check for `hal_efuse_macaddr_offset()` returning `-1` and not proceeding from there, so Condition 22 won't happen.

There is this other one, overlapping memory, that is being reported repeatedly in many parts of the tree:
```
os_dep/linux/ioctl_mp.c:1515:4:
  Type: Copy of overlapping memory (OVERLAPPING_COPY)

os_dep/linux/ioctl_mp.c:1499:2:
  1. path: Switch case value "MP_SINGLE_TONE_TX".
os_dep/linux/ioctl_mp.c:1514:3:
  2. path: Condition "bStartTest != 0", taking true branch.
os_dep/linux/ioctl_mp.c:1515:4:
  3. overlapping_copy: In the call to function "sprintf", the arguments "extra" and "extra" may point to the same object.
```
* I honestly don't think this is a big issue, but it might be worth taking a look how `sprintf` is being used to guarantee that no overlap is really happening.

Anyway, what we are trying to bring here are some inputs to work with, from a security perspective.
I really hope it can be useful.

Thanks!

[1] https://github.com/canonical/ubuntu-mir 
[2] https://cwe.mitre.org/data/definitions/457.html

**Comment by lwfinger** [comment]
I fixed the uninitialized array message shown by Coverity, and "fixed" a number of the overlapping string messages, mostly by turning off the MP (Manufacturing Process) stuff as it is only used by the factory to verify the chip. It is never used in the field.

I also checked the code with Smatch, which found one array overrun. That is also now fixed.

The Realtek factory drivers are a real hunk of crap, and likely have a lot of security flaws, but for the moment, they are all that is available for the RTW8723DS. The Linux driver for RTW8723DE was structured to handle interconnects other than PCIe, and the USB drivers for the rtw88 devices was just accepted into wireless-next. It will only be a matter of time before the SDIO driver is written, and this one can be ignored.
